sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -1882,3 +1882,311 @@ case class ArrayRepeat(left: Expression, right: Expression)
   }
 
 }
+
+object ArraySetLike {
+  val kindUnion = 1
+
+  private val MAX_ARRAY_LENGTH: Int = ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH
+
+  def toArrayDataInt(hs: OpenHashSet[Int]): ArrayData = {
+    val array = new Array[Int](hs.size)
+    var pos = hs.nextPos(0)
+    var i = 0
+    while (pos != OpenHashSet.INVALID_POS) {
+      array(i) = hs.getValue(pos)
+      pos = hs.nextPos(pos + 1)
+      i += 1
+    }
+
+    val numBytes = 4L * array.length
+    val unsafeArraySizeInBytes = UnsafeArrayData.calculateHeaderPortionInBytes(array.length) +
+      org.apache.spark.unsafe.array.ByteArrayMethods.roundNumberOfBytesToNearestWord(numBytes)
+    // Since UnsafeArrayData.fromPrimitiveArray() uses long[], max elements * 8 bytes can be used
+    if (unsafeArraySizeInBytes <= Integer.MAX_VALUE * 8) {
```

Comment:
`Integer.MAX_VALUE`?

Reply:
Maybe we should do the same calculation as in `UnsafeArrayData.fromPrimitiveArray()` to be clear what you are doing here.